src/disco/pack/fd_pack.c

```diff
@@ -0,0 +1,225 @@
+#include "fd_pack.h"
+
+#include "../mux/fd_mux.h"
+
+#define BLOCK_DURATION_NS (400UL*1000UL*1000UL)
+
+/* About 1.5 kB on the stack */
+#define FD_PACK_PACK_MAX_OUT (16UL)
+
+/* 1.5 M cost units, enough for 1 max size transaction */
+const ulong CUS_PER_MICROBLOCK = 1500000UL;
+
+const float VOTE_FRACTION = 0.75;
+
+typedef struct {
+  fd_pack_t *  pack;
+  fd_txn_p_t * cur_slot;
+
+  long block_duration_ticks;
+  long block_end;
+
+  fd_pack_in_ctx_t * in;
+
+  ulong    out_cnt;
+  ulong ** out_busy;
+
+  void * out_wksp;
+  ulong  out_chunk0;
+  ulong  out_wmark;
+  ulong  out_chunk;
+} fd_pack_ctx_t;
+
+static inline void
+before_poll( void * _ctx ) {
+  fd_pack_ctx_t * ctx = (fd_pack_ctx_t *)_ctx;
+
+  if( FD_UNLIKELY( ctx->cur_slot ) ) {
+    /* overrun while processing a slot, return it to the pool */
+    fd_pack_insert_txn_cancel( ctx->pack, ctx->cur_slot );
+    ctx->cur_slot = NULL;
+  }
```

Comment:
It took me a second to understand why this was here. How do you feel about adding an `overrun` callback? The init, fini/cancel pattern is fairly common in FD, so this would kind of echo it

Reply:
It's not a bad idea but I clarified the comment for now, I think we can add these callbacks when they are more strictly needed.